src/unistd.rs

```diff
@@ -2392,3 +2397,425 @@ pub fn access<P: ?Sized + NixPath>(path: &P, amode: AccessFlags) -> Result<()> {
+                None
+            }
+        } else {
+            Some(Err(Error::Sys(Errno::last())))
+        }
+    }
+}
+
+impl Queryable<GroupQuery> for Group {
+    fn query(q: GroupQuery) -> Option<Result<Self>> {
+        let mut cbuf = match q {
+            GroupQuery::GidWithBufsize(_, size) |
+            GroupQuery::NameWithBufsize(_, size) => {
+                vec![0 as c_char; size]
+            },
+            _ => {
+                vec![0 as c_char; PWGRP_BUFSIZE]
+            }
+        };
+
+        let mut grp: libc::group =  unsafe { mem::uninitialized() };
+        let mut res = ptr::null_mut();
+
+        let error = {
+            unsafe { Errno::clear(); }
+
+            match q {
+                GroupQuery::GidWithBufsize(Gid(gid), _) |
+                GroupQuery::Gid(Gid(gid)) => {
+                    unsafe {
+                        libc::getgrgid_r(gid, &mut grp,
+                                         cbuf.as_mut_ptr(),
+                                         cbuf.len(), &mut res)
+                    }
+                },
+                GroupQuery::NameWithBufsize(name, _) |
+                GroupQuery::Name(name) => {
+                    unsafe {
+                        libc::getgrnam_r(CString::new(name).unwrap().as_ptr(),
+                                         &mut grp, cbuf.as_mut_ptr(), cbuf.len(), &mut res)
+                    }
+                },
+            }
+
+        };
+
+        if error == 0 {
+            if !res.is_null() {
+                Some(Ok(Group::from(res)))
+            } else {
+                None
+            }
+        } else {
+            Some(Err(Error::Sys(Errno::last())))
+        }
+    }
+}
+
+#[cfg(not(any(target_os = "android",
+              target_os = "ios",
+              target_os = "macos",
+              target_env = "musl")))]
+mod usergroupiter {
+    use libc::{self, c_char};
+    use Result;
+    use errno::Errno;
+    use super::{Error, User, Group, PWGRP_BUFSIZE};
+    use std::{mem, ptr};
+
+    /// Used to get all of the users on the system.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use nix::unistd::Users;
+    /// Users::new()
+    ///     .map(|e|e.map(
+    ///         |pw| println!("{}\t{}",
+    ///                       pw.name,
+    ///                       pw.uid)))
+    ///     .collect::<Vec<_>>();
+    ///
+    /// ```
+    ///
+    /// # Safety
+    ///
+    /// This iterator should not be used in different threads without synchronization; while doing so
+    /// will not cause undefined behavior, because modern systems lack re-entrant versions of
+    /// `setpwent` and `endpwent`, it is very likely that iterators running in different threads will
+    /// yield different numbers of items.
+    #[derive(Debug, Clone, Eq, PartialEq, Hash, Default)]
+    pub struct Users(usize);
+
+    impl Users {
+        /// Create a new `Users` instance with default buffer size.
+        pub fn new() -> Self {
```

Comment:
I suggest combining `new` and `with_bufsize` into `fn new(bufsize: Option<usize>)`.

Reply:
I used an alternative option. I made:

```
    #[derive(Debug, Clone, Eq, PartialEq, Hash)]
    pub struct Users(usize);

    impl Default for Users {
        fn default() -> Self {
            Users::with_capacity(PWGRP_BUFSIZE)
        }
    }

    impl Users {
        /// Create a new `Users` instance with given capacity.
        pub fn with_capacity(bufsize: usize) -> Self {
            unsafe { libc::setpwent(); }
            Users(bufsize)
        }

        /// Get the buffer size this `Users` instance was created with.
        pub fn bufsize(&self) -> usize {
            self.0
        }
    }
```
What do you think?